src/optimizer/OptRule.cpp

```diff
@@ -69,10 +71,39 @@ StatusOr<MatchedResult> OptRule::match(const OptGroupNode *groupNode) const {
 }
 
 bool OptRule::match(const MatchedResult &matched) const {
-    UNUSED(matched);
-    // Return true if subclass doesn't override this interface,
-    // so optimizer will only check whether pattern is matched
-    return true;
+    return checkDataflowDeps(matched, matched.node->node()->outputVar(), true);
+}
+
+bool OptRule::checkDataflowDeps(const MatchedResult &matched,
+                                const std::string &var,
+                                bool isRoot) const {
+    auto node = matched.node;
+    auto planNode = node->node();
+    const auto &outVarName = planNode->outputVar();
+    if (outVarName != var) {
+        return false;
+    }
+    auto symTbl = planNode->qctx()->symTable();
+    auto outVar = symTbl->getVar(outVarName);
+    // Check whether this variable is read by multiple other plan nodes except the root
+    if (outVar->readBy.size() > 1 && !isRoot) {
+        return false;
+    }
+
+    if (matched.dependencies.empty()) {
+        return true;
+    }
+
+    DCHECK_EQ(matched.dependencies.size(), node->dependencies().size());
+
+    if (matched.dependencies.size() == 1U) {
```

Comment:
I think you could call it recursively in for loop directly instead of check size of dependencies.

Reply:
Good idea. I'll refactor the inputVar getter of PlanNode to do this.